File: bots/box-bot/MyBot.cpp
#include <stdlib.h>
#include <time.h>
#include <cstdlib>
#include <ctime>
#include <time.h>
#include <set>
#include <fstream>

#include "hlt.hpp"
#include "networking.hpp"

hlt::GameMap presentMap;
unsigned char myID;
int turn;

int findNearestEnemyDirection(hlt::Location loc) {
    unsigned char direction = NORTH;
    int maxDistance = presentMap.width / 2;
    int maxSearch = presentMap.width / 2;
    for (int d = 0; d < 5; d++) {
        int distance = 0;
        hlt::Location current = loc;
        hlt::Site site = presentMap.getSite(current, d);
        while (site.owner == myID && distance < maxDistance) {
            distance ++;
            current = presentMap.getLocation(current, d);
            site = presentMap.getSite(current);
        }

        float priority = 0;
        float away = 2;
        int travelled = 0;
        while (site.owner != myID && travelled < maxSearch) {
            travelled++;
            away = away * away;
            if (site.owner == 0) {
                priority -= (site.strength) / away;
            } else if (travelled < 1){
                priority += (site.strength) / away;
            }
            priority += (site.production) / away;
            current = presentMap.getLocation(current, d);
            site = presentMap.getSite(current);
        }

        priority = priority / 5;
        distance = distance - priority;

        if (distance < maxDistance) {
            direction = d;
            maxDistance = distance;
        }


    }
    return direction;
}

// int maxDir(int dir[5]) {
//     int max = 0;
//     for (int i = 0; i < 5; i ++) {
//         if (dir[i] > dir[max]) {
//             max = i;
//         }
//     }
//     return max;
// }

int priority(hlt::Site site) {
    return site.production - site.strength / 5;
}

int calcDamage(hlt::Location location) {
    int damage = 0;
    for (int i = 0; i < 5; i++) {
        if (presentMap.getSite(location, i).owner != 0
            && presentMap.getSite(location, i).owner != myID)
            damage += presentMap.getSite(location).strength;
    }
    return damage;
}

unsigned char assign_move(hlt::GameMap presentMap,
        unsigned char myID, unsigned short b, unsigned short a) {
    hlt::Site site = presentMap.getSite({ b, a });
    hlt::Location location;
    location.x = b;
    location.y = a;
    bool outer = false;
    unsigned char leastResistance = 0;
    int damage = 0;
    for (int j = 0; j < 5; j++) {
        if (presentMap.getSite(location, j).owner == 0) {
            if (leastResistance == 0) {
                leastResistance = j;
            } else {
                //expansion algorithm
                if (priority(presentMap.getSite(location, leastResistance))
                    < priority(presentMap.getSite(location, j))) {
                    damage = priority(presentMap.getSite(location, j));
                    leastResistance = j;
                }

                //combat algorithm
                if (calcDamage(presentMap.getLocation(location, j)) > 0) {
                    if (damage < calcDamage(presentMap.getLocation(location, j))) {
                        damage = calcDamage(presentMap.getLocation(location, j));
                        leastResistance = j;
                    }
                }

                //conquer algorithm

            }
            outer = true;
        }
    }


    // int damage = 0;

    // for (int j = 0; j < 5; j++) {
    //     if (presentMap.getSite(location, j).owner != myID) {
    //         if (leastResistance == 0) {
    //             leastResistance = j;
    //         } else {
    //             if (resistance(presentMap.getSite(location, leastResistance))
    //                 > resistance(presentMap.getSite(location, j))) {
    //                 leastResistance = j;
    //             }
    //         }
    //         outer = true;
    //     }
    // }

    unsigned char move;
    if (outer == false) {
        if (site.strength < 5 * site.production) {
            move = STILL;
        } else {
            move = findNearestEnemyDirection(location);

            // if (site.strength
            //     + presentMap.getSite(location, move).strength > 255) {
            //     move = STILL;
            // }
        }
    } else {
        if (presentMap.getSite(location, leastResistance).strength >= site.strength) {
            move = STILL;
        } else {
            move = leastResistance;
        }
    }
    return move;
}


int main() {
    srand(time(NULL));

    std::cout.sync_with_stdio(0);
    getInit(myID, presentMap);
    sendInit("MyC++Bot");

    std::set<hlt::Move> moves;

    while(true) {
        moves.clear();

        getFrame(presentMap);

        for(unsigned short a = 0; a < presentMap.height; a++) {
            for(unsigned short b = 0; b < presentMap.width; b++) {
                if (presentMap.getSite({ b, a }).owner == myID) {
                    moves.insert({ { b, a }, assign_move(presentMap, myID, b, a)});
                }
            }
        }
        turn++;
        sendFrame(moves);
    }

    return 0;
}

File: bots/thor-bot1.5/MyBot.java
import java.util.*;
import java.util.List;

public class MyBot {
    public static void main(String[] args) throws java.io.IOException {

        final InitPackage iPackage = Networking.getInit();
        final int myID = iPackage.myID;
        final GameMap gameMap = iPackage.map;
        int[][] dirs = {{1,0},
        {0,1},
        {-1,0},
        {0,-1}
        };
        Direction[] direcs = {
        Direction.EAST,
        Direction.SOUTH,
        Direction.WEST,
        Direction.NORTH,
        Direction.STILL
        };
        int numb = 0;
        for (int y = 0; y < gameMap.height; y++) {
            for (int x = 0; x < gameMap.width; x++) {
                numb+=gameMap.getLocation(x, y).getSite().strength;
            }
        }
        numb/=(gameMap.width*gameMap.height*5);
        int time = gameMap.width*gameMap.height/5;
        Networking.sendInit("MyJavaBot4");

        int myX = -1;
        int myY = -1;
        int count = 0;
        while(true) {
            count++;
            if(count%time==0)
                numb++;
            List<Move> moves = new ArrayList<Move>();
            
            Networking.updateFrame(gameMap);
            int[][] scores = new int[gameMap.width][gameMap.height];
            LinkedList<Integer[]> list = new LinkedList<>();
            for (int y = 0; y < gameMap.height; y++) {
                for (int x = 0; x < gameMap.width; x++) {
                    final Location location = gameMap.getLocation(x, y);
                    final Site site = location.getSite();
                    if(site.owner == myID) {
                        for(int i = 0; i < 4; i++)
                        {
                            Site side = getSite(gameMap,x,y,dirs[i]);
                            if(side.owner!=myID)
                                if(256-side.strength>scores[x][y])
                                    scores[x][y] = 256-side.strength;
                        }
                    }
                    else if(site.strength < numb)
                        list.add(new Integer[]{x,y});
                    if(scores[x][y]>0)
                    {
                        //list.add(new Integer[]{x,y});
                    }
                }
            }
            for (int y = 0; y < gameMap.height; y++) {
                for (int x = 0; x < gameMap.width; x++) {
                    final Location location = gameMap.getLocation(x, y);
                    final Site site = location.getSite();
                    if(site.owner == myID) {
                        if(scores[x][y]==0)
                        {
                            int[] vals = getMin(x,y,list,gameMap.height,gameMap.width);
                            if(site.strength > 1 && vals[0]+vals[1]+site.strength >= numb+5)
                            {
                                if(Math.abs(vals[0])>Math.abs(vals[1]))
                                {
                                    if(vals[0]>0)
                                        moves.add(new Move(location, Direction.EAST));
                                    else
                                        moves.add(new Move(location, Direction.WEST));
                                }
                                else
                                {
                                    if(vals[1]>0)
                                        moves.add(new Move(location, Direction.SOUTH));
                                    else
                                        moves.add(new Move(location, Direction.NORTH));
                                }
                            }
                            else
                                moves.add(new Move(location, Direction.STILL));
                        }
                        else
                        {
                            int dir = 4;
                            LinkedList<Integer> directs = new LinkedList<>();
                            if(256 - site.strength < scores[x][y])
                            {
                                for(int i = 0; i < 4; i++)
                                {
                                    Site side = getSite(gameMap,x,y,dirs[i]);
                                    if(side.owner!=myID)
                                        if(256-side.strength==scores[x][y])
                                            directs.add(i);
                                }
                                moves.add(new Move(location, direcs[directs.get((int)(Math.random()*directs.size()))]));
                            }
                            else
                                moves.add(new Move(location, Direction.STILL));
                        }
                    }
                }
            }
            Networking.sendFrame(moves);
        }
    }
    private static Site getSite(GameMap map, int x, int y, int... dir)
    {
        return map.getLocation((map.width+dir[0]+x)%map.width,(map.height+dir[1]+y)%map.height).getSite();
    }
    private static int[] getMin(int x, int y, LinkedList<Integer[]> list, int height, int width)
    {
        int[] res = {999999,999999};
        for(Integer[] i: list)
            if(mindist(i[0]-x,width)+mindist(i[1]-y,height) < Math.abs(res[0]) + Math.abs(res[1]))
            {
                res[0]=Math.abs(i[0]-x) < Math.abs(width-(i[0]-x))?i[0]-x:width-(i[0]-x);
                res[1]=Math.abs(i[1]-y) < Math.abs(height-(i[1]-y))?i[1]-y:height-(i[1]-y);
            }
        return res;
    }
    private static int mindist(int dif, int max)
    {
        dif = Math.abs(dif);
        return dif < max-dif?dif:max-dif;
    }
}
